core/nodejsActionBase/runner.js

```diff
@@ -42,29 +42,62 @@ function initializeActionHandler(message) {
+                const indexJSExists = fs.existsSync('index.js')
+                const indexMJSExists = fs.existsSync('index.mjs')
+                if (index === undefined && !packageJsonExists && !indexJSExists && !indexMJSExists) {
+                    return Promise.reject('Zipped actions must contain either package.json or index.[m]js at the root.');
+                }
+
+                let mainFile
+                if (index !== undefined) {
+                    // Backwards compat: We allow for main definitions like: `file.a.b.c`
+                    // where we'd import the function a.b.c from `file.[m]js`.
+                    if (fs.existsSync(index + '.js')) {
+                        mainFile = index + '.js'
+                    } else if (fs.existsSync(index + '.mjs')) {
+                        mainFile =  index + '.mjs'
+                    }
+                }
+                if (!mainFile && packageJsonExists) {
+                    // Infer the main file from package.json by default.
+                    let package = JSON.parse(fs.readFileSync('package.json'));
+                    mainFile = package.main
+                } 
+                if (!mainFile && indexMJSExists) {
+                    mainFile = 'index.mjs'
+                }
+                if (!mainFile) {
+                    mainFile = 'index.js'
                 }
 
                 //  The module to require.
-                let whatToRequire = index !== undefined ? path.join(moduleDir, index) : moduleDir;
-                let handler = eval('require("' + whatToRequire + '").' + main);
-                return assertMainIsFunction(handler, message.main);
+                let handler = eval('import("' + path.join(moduleDir, mainFile) + '").then(evaled => evaled.' + main + ')');
+                return handler.then(func => assertMainIsFunction(func, message.main))
             })
             .catch(error => Promise.reject(error));
     } else try {
-        let handler = eval(
-          `(function(){
-               ${message.code}
-               try {
-                 return ${message.main}
-               } catch (e) {
-                 if (e.name === 'ReferenceError') {
-                    return module.exports.${message.main} || exports.${message.main}
-                 } else throw e
-               }
-           })()`);
-        return assertMainIsFunction(handler, message.main);
+        // Write file as ES modules need to be loaded from files.
+        fs.writeFileSync('index.mjs', message.code)
+        let handler = eval('import("' + process.cwd() + '/index.mjs").then(evaled => evaled.' + message.main + ')');
+        return handler
+            .then(func => assertMainIsFunction(func, message.main))
+            .catch(_ => {
+                // If loading from file didn't work, we're dealing with a legacy function with no
+                // defined exports. Use "the old way" of evaling it.
+                let handler = eval(
```

Comment:
It might make sense to swap this logic around to avoid writing a file for the current common case. On the other hand, I think ES modules is the "correcter" way, so arguably it should be preferred.

Reply:
Had a similar thought. Deferring to your judgement.

Reply:
Went with swapping it around. Avoiding file ops is a good thing.